aiopg/connection.py

```diff
@@ -124,10 +142,26 @@ def __init__(self, dsn, loop, timeout, waiter, echo, **kwargs):
         self._notifies = asyncio.Queue(loop=loop)
         self._weakref = weakref.ref(self)
         self._loop.add_reader(self._fileno, self._ready, self._weakref)
+        self._conn_timeout_handler = None
+        if self._conn_timeout:
+            self._conn_timeout_handler = self._loop.call_later(
+                self._conn_timeout, self._shutdown, self._weakref)
 
         if loop.get_debug():
             self._source_traceback = traceback.extract_stack(sys._getframe(1))
 
+    @staticmethod
+    def _shutdown(weak_self):
+        # Make sure that we won't get stuck in a blocking read or write
+        # inside poll, then give libpq a chance to try another host.
+        # If there is an error, we'll get it from poll.
+        self = weak_self()
+
+        if self._conn.status == CONN_STATUS_CONNECTING:
```

Comment:
This occasionally fails due to `self` being `None`.

Minimal fix would be to skip the rest when that is the case,

but generally there shouldn't be active timers for a deleted object.